client/src/main/java/io/pravega/client/stream/impl/ControllerResolverFactory.java

```diff
@@ -137,50 +146,154 @@ public void start(Listener listener) {
+                        long scheduleDelay = 0;
+                        if (lastUpdateDuration < FAILURE_RETRY_TIMEOUT_MS) {
+                            scheduleDelay = FAILURE_RETRY_TIMEOUT_MS - lastUpdateDuration;
+                        }
+                        this.scheduledFuture = this.scheduledExecutor.schedule(
+                                this::getControllers, scheduleDelay, TimeUnit.MILLISECONDS);
+                    }
+                }
+            }
+        }
+
+        /**
+         * The controller discovery API invoker.
+         * This refreshes the list of controller addresses to be used by the gRPC transport.
+         * The discovery process will be rescheduled in the end after a delay which is calculated based on whether
+         * the controller addresses have been fetched successfully or not.
+         */
         private void getControllers() {
+            log.info("Attempting to refresh the controller server endpoints");
+            final ResolvedServerInfoGroup serverInfoGroup;
+            long nextScheduleTimeMS = REFRESH_INTERVAL_MS;
             try {
-                final ServerResponse controllerServerList =
-                        this.client.getControllerServerList(ServerRequest.getDefaultInstance());
-                final ResolvedServerInfoGroup serverInfoGroup = ResolvedServerInfoGroup.builder()
-                        .addAll(controllerServerList.getNodeURIList()
-                                .stream()
-                                .map(node ->
-                                        new ResolvedServerInfo(
-                                                new InetSocketAddress(node.getEndpoint(), node.getPort())))
-                                .collect(Collectors.toList()))
-                        .build();
+                if (this.enableDiscovery) {
+                    // Make an RPC call to the bootstrapped controller servers to fetch all active controllers.
+                    final ServerResponse controllerServerList =
+                            this.client.getControllerServerList(ServerRequest.getDefaultInstance());
+                    serverInfoGroup = ResolvedServerInfoGroup.builder()
+                            .addAll(controllerServerList.getNodeURIList()
+                                    .stream()
+                                    .map(node ->
+                                            new ResolvedServerInfo(
+                                                    new InetSocketAddress(node.getEndpoint(), node.getPort())))
+                                    .collect(Collectors.toList()))
+                            .build();
+                } else {
+                    // Resolve the bootstrapped server hostnames to get the set of controllers.
+                    final ArrayList<InetSocketAddress> resolvedAddresses = new ArrayList<>();
+                    this.bootstrapServers.forEach(address -> {
```

Comment:
(Unrelated to the other thread on this line)
bootstrapServers is guared by $lock but is not protected on this line. This method is effectively public because it is passed as a function externally. So no lock is held.